evm/db/shard.py

```diff
@@ -35,75 +36,102 @@ def __init__(self, db):
+                    shard_id,
+                    period,
+                )
+            )
+        else:
+            return canonical_hash
+
+    def get_canonical_header(self, shard_id: int, period: int) -> CollationHeader:
+        collation_hash = self.get_canonical_hash(shard_id, period)
+        return self.get_header_by_hash(collation_hash)
+
+    def get_canonical_body(self, shard_id: int, period: int) -> bytes:
+        header = self.get_canonical_header(shard_id, period)
+        return self.get_body_by_chunk_root(header.chunk_root)
+
+    def get_canonical_collation(self, shard_id: int, period: int) -> Collation:
+        collation_hash = self.get_canonical_hash(shard_id, period)
+        return self.get_collation_by_hash(collation_hash)
+
+    #
+    # Collation Setters
+    #
     def add_header(self, header: CollationHeader) -> None:
-        key = make_collation_header_lookup_key(header.shard_id, header.period)
-        self.db.set(key, rlp.encode(header))
+        self.db.set(header.hash, rlp.encode(header))
 
-    def add_body(self, shard_id: int, period: int, body: bytes) -> None:
-        key = make_collation_body_lookup_key(shard_id, period)
-        self.db.set(key, body)
-        self.mark_available(shard_id, period)
+    def add_body(self, body: bytes) -> None:
+        chunk_root = calc_chunk_root(body)
+        self.db.set(chunk_root, body)
+        self.mark_available(chunk_root)
 
     def add_collation(self, collation: Collation) -> None:
         self.add_header(collation.header)
-        self.add_body(collation.shard_id, collation.period, collation.body)
+        self.add_body(collation.body)
 
     #
     # Availability API
     #
-    def mark_unavailable(self, shard_id: int, period: int) -> None:
-        key = make_collation_availability_lookup_key(shard_id, period)
+    def mark_unavailable(self, chunk_root: bytes) -> None:
+        key = make_collation_availability_lookup_key(chunk_root)
         self.db.set(key, rlp.encode(False))
 
-    def mark_available(self, shard_id: int, period: int) -> None:
-        key = make_collation_availability_lookup_key(shard_id, period)
+    def mark_available(self, chunk_root: bytes) -> None:
+        key = make_collation_availability_lookup_key(chunk_root)
         self.db.set(key, rlp.encode(True))
 
-    def is_available(self, shard_id: int, period: int) -> bool:
-        key = make_collation_availability_lookup_key(shard_id, period)
+    def is_available(self, chunk_root: bytes) -> bool:
+        key = make_collation_availability_lookup_key(chunk_root)
         try:
             available = bool(rlp.decode(self.db.get(key), big_endian_int))
         except KeyError:
             return False
         return available
 
-    def is_unavailable(self, shard_id: int, period: int) -> bool:
-        key = make_collation_availability_lookup_key(shard_id, period)
+    def is_unavailable(self, chunk_root: bytes) -> bool:
```

Comment:
Just use `not self.is_available(chunk_root)`:
```python
def is_unavailable(self, chunk_root: bytes) -> bool:
    return not self.is_available(chunk_root)
```

Reply:
There's a slight difference: A collation can both be `not available` and `not unavailable`, but `availability_unknown`. I think this is important because we need to be able to distinguish between "I tried downloading the body but no one knew about it" and "I just got that header and haven't downloaded the body yet".

But I get that the names are confusing. Would it be better to have a single function `get_availability` that returns an enum `Availability` with values `available`, `unavailable` and `unknown`?

Reply:
Oh, I see. Agree with enum Availability. 👍

Reply:
You *could* get by with `True/False/None` here if you wan't to avoid the minor mental overhead of the enum, but I think the enum likely carries more semantic meaning so 👍